lib/usb/usb_f723.c

```diff
@@ -0,0 +1,485 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string.h>
+#include <libopencm3/cm3/common.h>
+#include <libopencm3/stm32/tools.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/usb/usbd.h>
+#include <libopencm3/usb/dwc/otg_hs.h>
+#include "usb_private.h"
+#include "usb_dwc_common.h"
+
+/* Receive FIFO size in 32-bit words. */
+#define RX_FIFO_SIZE 512
+
+#define MAX_BULK_PACKET_SIZE 512
+#define USB_ENDPOINT_COUNT 9
+
+static struct incoming_packet
+{
+	bool	is_packet_present;
+	int	packet_length;
+	uint8_t	packet_data[MAX_BULK_PACKET_SIZE];
+} stashed_packets[USB_ENDPOINT_COUNT];
+
+static struct _usbd_device _usbd_dev;
+
+/*
+ * Initialize the USB device controller hardware of the STM32.
+ *
+ * Note: this initialization code was compiled from the libopencm3 usb_f207.c
+ * source, and the st cube sources. Note that the st manuals state that some delays
+ * are necessary at certain places. This code works for usb hosts that I tested on
+ * without the delays, but this is bending the rules.
+ */
+static usbd_device *stm32f723_usbd_init(void)
+{
+	rcc_periph_clock_enable(RCC_OTGPHYC);
+	rcc_periph_clock_enable(RCC_OTGHSULPI);
+
+	rcc_periph_clock_enable(RCC_OTGHS);
+	OTG_HS_GINTSTS = OTG_GINTSTS_MMIS;
+
+	OTG_HS_GCCFG |= OTG_GCCFG_PHYHSEN;
+
+	OTG_HS_PHYC_LDO |= OTG_PHYC_LDO_DISABLE;
+	while (!(OTG_HS_PHYC_LDO & OTG_PHYC_LDO_STATUS))
+		continue;
+	OTG_HS_PHYC_PLL1 = OTG_PHYC_PLL1_SEL_25MHZ;
+	OTG_HS_PHYC_TUNE |= OTG_PHYC_TUNE_INCURREN |
+		OTG_PHYC_TUNE_INCURRINT |
+		OTG_PHYC_TUNE_HSDRVDCCUR |
+		OTG_PHYC_TUNE_HSDRVRFRED |
+		OTG_PHYC_TUNE_HSDRVCHKITRM_20_935MA;
+	OTG_HS_PHYC_PLL1 |= OTG_PHYC_PLL1_ENABLE;
+	/* 2ms Delay required to get internal phy clock stable */
+	for (volatile size_t i = 0; i < 200 * 1000; ++i)
+		continue;
+
+	/* Wait for AHB idle. */
+	while (!(OTG_HS_GRSTCTL & OTG_GRSTCTL_AHBIDL))
+		continue;
+	/* Do core soft reset. */
+	OTG_HS_GRSTCTL |= OTG_GRSTCTL_CSRST;
+	while (OTG_HS_GRSTCTL & OTG_GRSTCTL_CSRST)
+		continue;
+
+	/* Force peripheral only mode. */
+	OTG_HS_GUSBCFG |= OTG_GUSBCFG_FDMOD | OTG_GUSBCFG_TRDT_MASK;
+
+	/* Restart the PHY clock. */
+	OTG_HS_PCGCCTL = 0;
+
+	OTG_HS_GRXFSIZ = stm32f723_usb_driver.rx_fifo_size;
+	_usbd_dev.fifo_mem_top = stm32f723_usb_driver.rx_fifo_size;
+
+	OTG_HS_DCTL |= OTG_DCTL_SDIS;
+
+	/* Unmask interrupts for TX and RX. */
+	OTG_HS_GAHBCFG |= OTG_GAHBCFG_GINT;
+	OTG_HS_GINTMSK = OTG_GINTMSK_ENUMDNEM | OTG_GINTMSK_RXFLVLM | OTG_GINTMSK_IEPINT
+		| OTG_GINTMSK_OEPINT | OTG_GINTMSK_USBSUSPM | OTG_GINTMSK_WUIM;
+
+	OTG_HS_DAINTMSK = 0xFFFFFFFF;
+
+	OTG_HS_DOEPMSK |= OTG_DOEPMSK_STUPM | OTG_DOEPMSK_XFRCM;
+	OTG_HS_DIEPMSK |= OTG_DIEPMSK_XFRCM;
+
+	OTG_HS_DCTL &= ~OTG_DCTL_SDIS;
+
+	return &_usbd_dev;
+}
+
+/* Note: the original 'usb_dwc_common.c' source code
```

Comment:
This comment should belong to the `stashed_packets` array variable, at lines 36-41